presto-memory-context/src/main/java/com/facebook/presto/memory/MemoryTrackingContext.java

```diff
@@ -0,0 +1,182 @@
+ *
+ * To perform local allocations clients should call localUserMemoryContext()/localSystemMemoryContext()
+ * and get a reference to the local memory contexts. The other child-originated allocations will go through
+ * reserveUserMemory()/reserveRevocableMemory()/setSystemMemory() methods.
+ */
+@ThreadSafe
+public class MemoryTrackingContext
+{
+    private final AggregatedMemoryContext userReservedAggregateMemoryContext;
+    private final AggregatedMemoryContext userRevocableAggregateMemoryContext;
+    private final AggregatedMemoryContext systemReservedAggregateMemoryContext;
+
+    private final LocalMemoryContext userReservedLocalMemoryContext;
+    private final LocalMemoryContext userRevocableLocalMemoryContext;
+    private final LocalMemoryContext systemReservedLocalMemoryContext;
+
+    public MemoryTrackingContext(
+            AggregatedMemoryContext userReservedAggregateMemoryContext,
+            AggregatedMemoryContext userRevocableAggregateMemoryContext,
+            AggregatedMemoryContext systemReservedAggregateMemoryContext)
+    {
+        this.userReservedAggregateMemoryContext = requireNonNull(userReservedAggregateMemoryContext, "userReservedAggregateMemoryContext is null");
+        this.userRevocableAggregateMemoryContext = requireNonNull(userRevocableAggregateMemoryContext, "userRevocableAggregateMemoryContext is null");
+        this.systemReservedAggregateMemoryContext = requireNonNull(systemReservedAggregateMemoryContext, "systemReservedAggregateMemoryContext is null");
+        this.userReservedLocalMemoryContext = userReservedAggregateMemoryContext.newLocalMemoryContext();
+        this.userRevocableLocalMemoryContext = userRevocableAggregateMemoryContext.newLocalMemoryContext();
+        this.systemReservedLocalMemoryContext = systemReservedAggregateMemoryContext.newLocalMemoryContext();
+    }
+
+    // below methods are for reserving memory locally
+    public void reserveUserMemory(long delta)
+    {
+        checkArgument(delta >= 0, "delta is negative");
+        userReservedLocalMemoryContext.addBytes(delta);
+    }
+
+    public void reserveRevocableMemory(long delta)
+    {
+        checkArgument(delta >= 0, "delta is negative");
+        userRevocableLocalMemoryContext.addBytes(delta);
+    }
+
+    public void setSystemMemory(long systemMemoryUsage)
+    {
+        checkArgument(systemMemoryUsage >= 0, "systemMemoryUsage is negative");
+        systemReservedLocalMemoryContext.setBytes(systemMemoryUsage);
+    }
+
+    // "free" methods always free from the local context, which is reflected to the aggregate context
+    public void freeUserMemory(long delta)
+    {
+        checkArgument(delta >= 0, "delta is negative");
+        checkArgument(delta <= userReservedLocalMemoryContext.getBytes(), "cannot free more memory than reserved");
+        userReservedLocalMemoryContext.addBytes(-delta);
+    }
+
+    public void freeRevocableMemory(long delta)
```

Comment:
All functions that take a `delta` will have to be removed as a result of my comments for the first commit.